### PR TITLE
Cluster wide quota stats via Helix

### PR DESCRIPTION
**Pull request by xunyin8**
Enables Cluster wide quota stats aggregation via Helix's Health Report and Task Framework.

1. Each node (instance in Helix) exposes node level quota stats via Helix Health Report.
2. The cluster wide aggregation will be delegated to a chosen live node via Helix task framework and `HelixAggregateTask`.

Node level stats will be reported by Helix's existing Health Report framework. The implementation of `QuotaHealthReport` will be responsible of exposing node level quota stats at every node when Helix is collecting health report for every instance.

Each node (instance) will be registered with the `HelixAggregateTask` which can be triggered to perform cluster wide aggregation. Every node is capable of performing cluster wide aggregation. The logic for performing the cluster wide aggregation are within `HelixClusterAggregator`. Trigger for the task can either be on demand or periodically. A generic job (job without specifying a target) will pick a live node (instance) in the cluster and trigger its `HelixAggregateTask` which will perform the cluster wide aggregation and store the final result in Helix's PropertyStore.

More details about the design and implementation can be found [here](https://docs.google.com/a/linkedin.com/document/d/1qz7d-Dp6FQkhZUd6LrAinZZnnP5nn66LU5hjPIEUpDI/edit?usp=sharing).

Code/workflow tested locally with unit tests and running a local mock cluster.

Unit test coverage:

| Class          | Method, %     | Line, %        |
|----------------|---------------|----------------|
| HelixClusterAggregator | 100% (5/ 5) | 100% (57/ 57) |

Code style/format applied.
`./gradlew build && ./gradlew test` success.

### CODE REVIEW

**Comment by xunyin8** [comment]
@pnarayanan I added Javadoc to most of them but I think missed a few. Thanks for pointing out I will add them.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java [hunk 1]
```diff
@@ -59,16 +72,40 @@
...
+  public void initialize(String hostName, int port, List<HealthReport> healthReports) throws IOException {
     String instanceName = ClusterMapUtils.getInstanceName(hostName, port);
     manager = helixFactory.getZKHelixManager(clusterName, instanceName, InstanceType.PARTICIPANT, zkConnectStr);
     StateMachineEngine stateMachineEngine = manager.getStateMachineEngine();
     stateMachineEngine.registerStateModelFactory(LeaderStandbySMD.name, new AmbryStateModelFactory());
+    Map<String, TaskFactory> taskFactoryMap = new HashMap<>();
+    for (HealthReport healthReport : healthReports) {
+      if (healthReport.getAggregatePeriodInMinutes() != Utils.Infinite_Time) {
+        // enable cluster wide aggregation for the health report
+        final HelixClusterAggregator aggregator =
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java]: what is the necessity to instantiate HelixClusterAggregator within HelixParticipant. Why not keep the logic in HelixParticipant clean. Just have the task registration alone here. And move the instantiation of HelixClusterAggregator to HelixAggregateTask

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java [hunk 2]
```diff
@@ -14,18 +14,31 @@
...
+import org.apache.helix.task.TaskFactory;
+import org.apache.helix.task.TaskStateModelFactory;
 import org.json.JSONException;
 
 
 /**
  * An implementation of {@link ClusterParticipant} that registers registers as a participant to a Helix cluster.
  */
 class HelixParticipant implements ClusterParticipant {
+  private static long WORKFLOW_EXPIRY = TimeUnit.SECONDS.toMillis(30);
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java]: what is the purpose of this. And I don't see it being used anywhere

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java (hunk 1 above)
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java]: Also move the task registration to a separate method. Let the initialize be simple for readability purposes. 

```
 stateMachineEngine.registerStateModelFactory(LeaderStandbySMD.name, new AmbryStateModelFactory());
 registerHealthReportTasks(healthReports);
    try {
      manager.connect();
    }
.
.
.
```

File: ambry-api/src/main/java/com.github.ambry/server/HealthReport.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Health report for reporting various stats within a node.
+ */
+public interface HealthReport {
+  /**
+   * Get the field name (accessor) for the stats in the health report.
+   * @return the key in the health report that holds the stats
+   */
+  String getFieldName();
```
Comment (nsivabalan) [inline ambry-api/src/main/java/com.github.ambry/server/HealthReport.java]: getSimpleFieldNameForQuotaStats()

File: ambry-server/src/main/java/com.github.ambry.server/StatsManager.java [hunk 4]
```diff
@@ -134,7 +135,7 @@ void collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partition
       try {
         long fetchAndAggregatePerStoreStartTimeMs = time.milliseconds();
         StatsSnapshot statsSnapshot = store.getStoreStats().getStatsSnapshot();
-        aggregate(aggregatedSnapshot, statsSnapshot);
+        StatsSnapshot.aggregate(aggregatedSnapshot, statsSnapshot);
```
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/StatsManager.java]: will you put up a follow up PR to remove the scheduler within StatsManager which aggregates and publishes the stats locally

File: ambry-server/src/main/java/com.github.ambry.server/QuotaHealthReport.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Map;
+import org.apache.helix.healthcheck.HealthReportProvider;
+
+
+/**
+ * Customized Helix Health Report for collecting quota stats per node (instance).
+ */
+class QuotaHealthReport extends HealthReportProvider implements HealthReport {
+  private static final String REPORT_NAME = "QuotaReport";
+  private static final String FIELD_NAME = "QuotaStats";
```
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/QuotaHealthReport.java]: quotaFieldName

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixAggregateTask.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      List<String> instanceNames = manager.getClusterManagmentTool().getInstancesInCluster(manager.getClusterName());
+      List<String> statsWrappersJSON = new ArrayList<>();
+      for (String instanceName : instanceNames) {
+        PropertyKey.Builder keyBuilder = helixDataAccessor.keyBuilder();
+        HelixProperty record = helixDataAccessor.getProperty(keyBuilder.healthReport(instanceName, healthReportId));
+        if (record != null && record.getRecord() != null) {
+          statsWrappersJSON.add(record.getRecord().getSimpleField(fieldName));
+        }
+      }
+      Pair<String, String> results = clusterAggregator.doWork(statsWrappersJSON);
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixAggregateTask.java]: why ClusterAggregator needs to a separate file in itself. Why not internal class? It is being used only by this task right

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixAggregateTask.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+      String resultId = String.format("Aggregated_%s", healthReportId);
+      ZNRecord znRecord = new ZNRecord(resultId);
+      znRecord.setSimpleField(RAW_VALID_SIZE_FIELD_NAME, results.getFirst());
+      znRecord.setSimpleField(VALID_SIZE_FIELD_NAME, results.getSecond());
+      String path = String.format("/%s", resultId);
+      manager.getHelixPropertyStore().set(path, znRecord, 1);
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixAggregateTask.java]: what does the 3rd arg mean.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    }
+    rawBaseSnapshot.setValue(totalValue);
+  }
+
+  /**
+   * Aggregate the given {@link StatsSnapshot} with the base {@link StatsSnapshot} by partition with the following
+   * rules:
+   * 1. Append the partition entry if the base doesn't already contain an entry for a given partition.
+   * 2. If the base already has an entry for a given partition, update/replace the entry only if:
+   *    a) Timestamp difference of a given partition entry and base entry are within one aggregation period, and the
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java]: lets discuss this offline tomorrow. I need some clarifications.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    for (String statsWrapperJSON : statsWrappersJSON) {
+      if (statsWrapperJSON != null) {
+        StatsWrapper snapshotWrapper = mapper.readValue(statsWrapperJSON, StatsWrapper.class);
+        StatsWrapper snapshotWrapperCopy = mapper.readValue(statsWrapperJSON, StatsWrapper.class);
+        combineRaw(rawPartitionSnapshot, snapshotWrapper);
+        combine(partitionSnapshot, snapshotWrapperCopy, partitionTimestampMap);
+      }
+    }
+    StatsSnapshot reducedRawSnapshot = reduce(rawPartitionSnapshot);
+    StatsSnapshot reducedSnapshot = reduce(partitionSnapshot);
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java]: what is the format of this reduced snapshot? Does it have partition info or is it removed?

File: ambry-tools/src/main/java/com.github.ambry/clustermap/HelixClusterWideAggregationTool.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    List<TaskConfig> taskConfigs = new ArrayList<>();
+    taskConfigs.add(new TaskConfig.Builder().setTaskId(TASK_ID)
+        .setCommand(String.format("%s_%s", HelixAggregateTask.TASK_COMMAND_PREFIX, REPORT_NAME))
+        .build());
+    jobConfigBuilder.addTaskConfigs(taskConfigs);
+    workflowBuilder.addJob(JOB_ID, jobConfigBuilder);
+    Workflow workflow = workflowBuilder.build();
+    taskDriver.start(workflow);
+    String jobName = String.format("%s_%s", workflowName, JOB_ID);
+    taskDriver.pollForJobState(workflowName, jobName, taskTimeoutInSecs, TaskState.COMPLETED);
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/clustermap/HelixClusterWideAggregationTool.java]: don't need to wait until the job is complete? This will not return at all. May you can print the jobId info. So that user can check the status of the job in zookeeper.

File: ambry-tools/src/main/java/com.github.ambry/clustermap/HelixClusterWideAggregationTool.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    OptionSet options = parser.parse(args);
+    String zkAddress = options.valueOf(zkAddressOpt);
+    String clusterName = options.valueOf(clusterNameOpt);
+    String workflowName = options.valueOf(workflowNameOpt);
+    Long taskTimeoutInSecs = options.valueOf(taskTimeoutInSecsOpt);
+    ZkClient zkClient = new ZkClient(zkAddress, SESSION_TIMEOUT, CONNECTION_TIMEOUT, new ZNRecordSerializer());
+    TaskDriver taskDriver = new TaskDriver(zkClient, clusterName);
+    Workflow.Builder workflowBuilder = new Workflow.Builder(workflowName);
+    workflowBuilder.setExpiry(WORKFLOW_EXPIRY);
```
Comment (nsivabalan) [inline ambry-tools/src/main/java/com.github.ambry/clustermap/HelixClusterWideAggregationTool.java]: what is the purpose of this config ?

File: ambry-api/src/main/java/com.github.ambry/server/HealthReport.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.github.ambry.server;
+
+/**
+ * Health report for reporting various stats within a node.
+ */
+public interface HealthReport {
```
Comment (xunyin8) [inline ambry-api/src/main/java/com.github.ambry/server/HealthReport.java]: I agree changing the name to AmbryHealthReport. However, if we implement HealthReportProvider (helix library) then we are forcing the dependency on Helix even for a static cluster. Is that ok?

File: ambry-api/src/main/java/com.github.ambry/server/HealthReport.java (hunk 3 above)
Comment (xunyin8) [inline ambry-api/src/main/java/com.github.ambry/server/HealthReport.java]: The HealthReport can be used to track more than just QuotaStats.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixAggregateTask.java (hunk 6 above)
Comment (xunyin8) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixAggregateTask.java]: This was done to abstract the cluster wide aggregation logic away from Helix related classes.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixAggregateTask.java (hunk 7 above)
Comment (xunyin8) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixAggregateTask.java]: > Set the type of ZNode see the valid values in AccessOption
Here are the values in AccessOption:
PERSISTENT = 0x1;
EPHEMERAL = 0x2;
PERSISTENT_SEQUENTIAL = 0x4;
EPHEMERAL_SEQUENTIAL = 0x8;
THROW_EXCEPTION_IFNOTEXIST = 0x10;

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java (hunk 2 above)
Comment (xunyin8) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java]: Leftover code from removing workflow triggers from HelixParticipant. Will remove.

File: ambry-server/src/main/java/com.github.ambry.server/StatsManager.java (hunk 4 above)
Comment (xunyin8) [inline ambry-server/src/main/java/com.github.ambry.server/StatsManager.java]: At the moment all changes are made to keep the old features functional. If you would like too, i can also just remove it.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java (hunk 9 above)
Comment (xunyin8) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java]: partition info is removed.

File: ambry-tools/src/main/java/com.github.ambry/clustermap/HelixClusterWideAggregationTool.java (hunk 11 above)
Comment (xunyin8) [inline ambry-tools/src/main/java/com.github.ambry/clustermap/HelixClusterWideAggregationTool.java]: I couldn't find a detailed description of this config. By default is set to 1 day `24 * 60 * 60 * 1000`. Perhaps we could just use the default.

File: ambry-api/src/main/java/com.github.ambry/config/ServerConfig.java [hunk 13]
```diff
@@ -34,13 +34,20 @@
   public final int serverSchedulerNumOfthreads;
 
   /**
-   * The option to enable or disable publishing stats via Health Report.
+   * The option to enable or disable publishing stats locally.
```
Comment (nsivabalan) [inline ambry-api/src/main/java/com.github.ambry/config/ServerConfig.java]: can you rename this to "server.stats.local.publish.enabled"

File: ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java [hunk 14]
```diff
@@ -163,10 +163,12 @@ public void startup() throws InstantiationException {
         partitionIds.add(replicaId.getPartitionId());
       }
       statsManager = new StatsManager(storageManager, partitionIds, registry, statsConfig, time);
-      statsManager.start();
+      if (serverConfig.serverStatsPublishEnabled) {
```
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java]: shouldn't the if block span lines 160 to 168?

File: ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java [hunk 15]
```diff
@@ -157,16 +157,22 @@ public void startup() throws InstantiationException {
           networkServer.getRequestResponseChannel(), requests);
       networkServer.start();
 
+      logger.info("Creating StatsManager to publish stats");
```
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java]: why not avoid creating StatsManager if serverConfig.serverStatsPublishEnabled is not set?

File: ambry-server/src/main/java/com.github.ambry.server/QuotaHealthReport.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+package com.github.ambry.server;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.helix.healthcheck.HealthReportProvider;
+
+
+/**
+ * Customized Helix Health Report for collecting quota stats per node (instance).
```
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/QuotaHealthReport.java]: some details about how the report would look like (structurally) would be nice.

File: ambry-server/src/main/java/com.github.ambry.server/StatsManager.java (hunk 4 above)
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/StatsManager.java]: sure. Don't need to remove the local publishing for now. Let it be

File: ambry-server/src/main/java/com.github.ambry.server/StatsManager.java [hunk 17]
```diff
@@ -143,22 +144,59 @@ void collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partition
...
-          baseSnapshot.getSubMap().put(entry.getKey(), new StatsSnapshot(0L, null));
+  StatsSnapshot fetchSnapshot(PartitionId partitionId, List<String> unreachableStores) {
+    StatsSnapshot statsSnapshot = null;
+    Store store = storageManager.getStore(partitionId);
+    if (store == null) {
+      unreachableStores.add(partitionId.toString());
+    } else {
+      try {
+        statsSnapshot = store.getStoreStats().getStatsSnapshot();
+      } catch (StoreException e) {
```
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/StatsManager.java]: log the exception

File: ambry-server/src/main/java/com.github.ambry.server/StatsManager.java [hunk 18]
```diff
@@ -143,22 +144,59 @@ void collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partition
...
+        unreachableStores.add(partitionId.toString());
+      }
+    }
+    return statsSnapshot;
+  }
+
+  /**
+   * Get the combined {@link StatsSnapshot} of all partitions in this node.
+   * @return a combined {@link StatsSnapshot} of this node
```
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/StatsManager.java]: some example on how a combined would look like. Easy to get confused that we might reduce it to just one map w/o per partition info

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.helix.HelixManager;
+import org.apache.helix.HelixProperty;
+import org.apache.helix.PropertyKey;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.task.Task;
+import org.apache.helix.task.TaskCallbackContext;
+import org.apache.helix.task.TaskResult;
+import org.apache.helix.task.UserContentStore;
+
+
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java]: java docs

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+class HelixHealthReportAggregatorTask extends UserContentStore implements Task {
+  public static final String TASK_COMMAND_PREFIX = "aggregate";
+  private static final String RAW_VALID_SIZE_FIELD_NAME = "raw_valid_data_size";
+  private static final String VALID_SIZE_FIELD_NAME = "valid_data_size";
+  private static final String TIMESTAMP_FIELD_NAME = "timestamp";
+  private final HelixManager manager;
+  private final HelixClusterAggregator clusterAggregator;
+  private final String healthReportId;
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java]: why switching from "healthReportName" to "healthReportId"

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+      List<String> statsWrappersJSON = new ArrayList<>();
+      for (String instanceName : instanceNames) {
+        PropertyKey.Builder keyBuilder = helixDataAccessor.keyBuilder();
+        HelixProperty record = helixDataAccessor.getProperty(keyBuilder.healthReport(instanceName, healthReportId));
+        if (record != null && record.getRecord() != null) {
+          statsWrappersJSON.add(record.getRecord().getSimpleField(fieldName));
+        }
+      }
+      Pair<String, String> results = clusterAggregator.doWork(statsWrappersJSON);
+      String resultId = String.format("Aggregated_%s", healthReportId);
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java]: can we make this path (helix property store path) a config and not generate it on the fly ?

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+      ZNRecord znRecord = new ZNRecord(resultId);
+      znRecord.setSimpleField(RAW_VALID_SIZE_FIELD_NAME, results.getFirst());
+      znRecord.setSimpleField(VALID_SIZE_FIELD_NAME, results.getSecond());
+      znRecord.setSimpleField(TIMESTAMP_FIELD_NAME, String.valueOf(SystemTime.getInstance().milliseconds()));
+      String path = String.format("/%s", resultId);
+      manager.getHelixPropertyStore().set(path, znRecord, AccessOption.PERSISTENT);
+      return new TaskResult(TaskResult.Status.COMPLETED, "Aggregation success");
+    } catch (Exception e) {
+      return new TaskResult(TaskResult.Status.FAILED, "Exception thrown");
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java]: can you log the stack trace from the exception

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+    relevantTimePeriodInMs = TimeUnit.MINUTES.toMillis(relevantTimePeriodInMinutes);
+    mapper.setVisibilityChecker(mapper.getVisibilityChecker().withFieldVisibility(JsonAutoDetect.Visibility.ANY));
+  }
+
+  /**
+   * Take a {@link List} of JSON string representation of {@link StatsWrapper} objects and perform cluster wide
+   * aggregation with them.
+   * @param statsWrappersJSON a {@link List} of JSON string representation of {@link StatsWrapper} objects from the
+   *                          node level
+   * @return a {@link Pair} of Strings whose first element is the raw (sum) aggregated stats and whose second element is
```
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java]: can you rephrase what the 2nd value is. Guess its not conveyed well.

**Comment by nsivabalan** [comment]
Merging the patch for now. Someone will pick this up to work on it later. Have ensured that this patch will not break anything as health reports are disabled as per configs and cluster wide aggregation will have to be triggered via tools.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java (hunk 15 above)
Comment (nsivabalan) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java]: StatManager initialization is required if serverConfig.serverStatsHealthReportEnabled is true.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java (hunk 21 above)
Comment (nsivabalan) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixHealthReportAggregatorTask.java]: Looks like we have to change 3 to 4 classes to bring in this field value from config. Leaving it as is for now.
